scrapy/downloadermiddlewares/httpcompression.py

```diff
@@ -27,7 +27,17 @@ def process_response(self, request, response, spider):
             content_encoding = response.headers.getlist('Content-Encoding')
             if content_encoding and not is_gzipped(response):
                 encoding = content_encoding.pop()
-                decoded_body = self._decode(response.body, encoding.lower())
+                try:
+                    decoded_body = self._decode(response.body, encoding.lower())
+                except (IOError, zlib.error):
+                    # Propagate decompression failures of successful responses.
+                    if 200 <= response.status < 300:
+                        raise
```

Comment:
What do you think about moving this logic to redirect middleware? It may catch these exception from HttpCompressionMiddleware and bypass them explicitly.